07-routing-gossip.md

```diff
@@ -445,10 +445,12 @@ or `node_id_2` otherwise.
 ### Requirements
 
 The origin node:
-  - MUST NOT send a created `channel_update` before `funding_locked` has been received.
+  - MUST NOT send a created `channel_update` before `channel_ready` has been received.
   - MAY create a `channel_update` to communicate the channel parameters to the
   channel peer, even though the channel has not yet been announced (i.e. the
   `announce_channel` bit was not set).
+    - MUST set the `short_channel_id` to either an `alias` it has
+	  received from the peer, or the real channel `short_channel_id`.
```

Comment:
> That's really confusing I think. In your example below: if B is the alias set by Bob for Bob->Alice, then B is what Bob should use in the channel_update for Bob->Alice. Otherwise the same alias points to one direction (in the routing hint) and the opposite direction (in a payment error)

Its confusing in that you have to generate channel updates differently based on the context, but you already likely have to do that to handle the different cases for onions now anyway - using real scids for public channels and refusing to do so for private ones.

> Curious how you have implemented it @TheBlueMatt? Did you use the sender or receiver alias for channel_update?

The receiver's alias makes sense here - the receiver isn't otherwise required to understand the sender's alias at all, and moreso its an optional TLV so we have to assume they don't know how to interpret it at all.

Reply:
Right, but the receiver doesn't have to use the alias in their invoices at all, if they don't want to (without the scid_privacy channel type set). This is, again, consistent with the SCID alias being an optional TLV, with no feature bit assigned to describing its presence.

Reply:
So, we are optimizing for the case where both the sender and the receiver understand aliases, but neither of `option_scid_privacy` or `option_zeroconf` are enabled, and the alias isn't used for anything except in the `channel_update`, and in a way that creates the confusion I layed out above? I don't know why we would want that.

In my opinion we should instead:
- if `option_scid_privacy` or `option_zeroconf` is enabled: have each side always use their own alias for their own `channel_update`
- otherwise: use the real `scid` for `channel_update`s.

This too would be consistent with the alias being an optional TLV: in the latter case the sender is always free to set the alias, and the receiver may or may not use it in their invoice.

Reply:
For public zero-conf channels, you'd also want to exchange the real SCID in updates

Reply:
> So, we are optimizing for the case 

I wouldn't say we're "optimizing" for any case here - this is the only way it can work. Otherwise what value do you put in the channel_update - you sent a SCID alias, you don't know if the peer supports it or remembered it at all, you cannot, there is no feature to indicate the peer must support it, there is no ACK of the scid alias you sent. They may have sent their channel_ready before you, so you can't use that as an ACK that they stored your SCID alias either.

> if option_scid_privacy or option_zeroconf is enabled:

Its not crazy to want to use SCID aliases even if the privacy extension is not negotiated on the channel, eg for backwards compat many LDK clients will do this today.

Reply:
>  what value do you put in the channel_update [...], you don't know if the peer supports it or remembered it at all

If `option_scid_privacy` or `option_zeroconf` is negotiated, you use the scid alias in the `channel_update` because you know that the peer supports it and they must remember it. Otherwise, you use the real scid just like before this PR, and the peer still has the option to use your alias in their invoice just like you wish.

> there is no ACK of the scid alias you sent

No ACK is required if the alias is used only when `option_scid_privacy` and/or `option_zeroconf` have been negotiated. You only have this ACK problem if you don't know whether the peer supports it, which is why we should simply use the real scid in that case, instead of sometimes using the sender's alias, sometimes the receiver's alias for the same `channel_update`.

> Its not crazy to want to use SCID aliases even if the privacy extension is not negotiated on the channel

I agree, and again, what I propose do not prevent that.

> For public zero-conf channels, you'd also want to exchange the real SCID in updates

Indeed: we would use the scid alias until the funding tx is confirmed, and the real scid after that.

Reply:
> If option_scid_privacy or option_zeroconf is negotiated, you use the scid alias in the channel_update because you know that the peer supports it and they must remember it.

Or we could not add any branching based on feature flags, no complicated logic to figure out which one to use and just keep it as-is :). I'm not really sure why you're advocating this - yea, its a bit awkward from a theoretical perspective, but the way its written now leaves the code much, much simpler.

Reply:
There will be branching anyway: either based on feature flags (I'd argue they are built for that purpose), or based on context (depending on what the `channel_update` is built for). Anyway I think both positions are pretty clear let's see what others think 🤓.

Reply:
Oh, apologies, using only "option_scid_privacy or option_zeroconf" doesn't work - you're totally allowed to start using a channel at 0conf even without `option_zeroconf`. This is in fact an important use-case because I believe only lnd is gonna bother using `option_zeroconf`, others can/will simply send a channel_ready without any confirmations.

Reply:
> which allows each side to tell whether the other side understand aliases

Not currently, you'd be adding an assumption that if a peer sent you an alias they also stored the alias you sent them, which is a new assumption that doesn't seem like it needs to make?

> It is still strictly simpler than using the receiver's scid alias, which requires the same branching + makes the scid context-dependent for the channel_update.

This doesn't match up with our implementation experience at all. On the sending side we just do `chan.get_short_channel_id().or(chan.latest_inbound_scid_alias())` which is nice and clean. On the receiving side, we use the same map we use for outbound relay - we already (obviously) have to have some listing of SCIDs that includes real and outbound-aliases in one place, we currently do not anywhere have a listing of real and inbound-alias SCIDs anywhere. Your proposed change would mandate we store a new hashmap for SCID aliases for the inbound side.